File: solutions/169_majority_element.rb
=begin
Given an array of size n, find the majority element. The majority element is the element that appears more than ⌊ n/2 ⌋ times.

You may assume that the array is non-empty and the majority element always exist in the array.

Example 1:

Input: [3,2,3]
Output: 3
Example 2:

Input: [2,2,1,1,1,2,2]
Output: 2
=end

# @param {Integer[]} nums
# @return {Integer}
def majority_element(nums)
    n = nums.length
    count_by_num = {}
    nums.each do |num|
        count_by_num[num] ||= 0
        count_by_num[num] += 1
        if count_by_num[num] > n/2
            return num
        end
    end
end

File: 169_majority_element.rs
/*
Given an array nums of size n, return the majority element.

The majority element is the element that appears more than ⌊n / 2⌋ times. You may assume that the majority element always exists in the array.

 

Example 1:

Input: nums = [3,2,3]
Output: 3
Example 2:

Input: nums = [2,2,1,1,1,2,2]
Output: 2
 

Constraints:

n == nums.length
1 <= n <= 5 * 104
-231 <= nums[i] <= 231 - 1
 

Follow-up: Could you solve the problem in linear time and in O(1) space?
*/


impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        let mut count = 1;
        let mut candidate = nums[0];
        for &n in &nums[1..] {
            if count == 0 {
                candidate = n;
            }
            if n == candidate {
                count += 1;
            } else {
                count -= 1;
            }
        }
        candidate
    }
}

